logstash-core/lib/logstash/instrument/collector.rb

```diff
@@ -1,79 +1,100 @@
+require "logstash/instrument/metric_store"
 require "logstash/util/loggable"
-require "concurrent/map"
+require "concurrent/timer_task"
 require "observer"
 require "singleton"
 require "thread"
 
 module LogStash module Instrument
+  # The Collector singleton is the single point of reference for all
+  # the metrics collection inside logstash, the metrics library will make
+  # direct calls to this class.
+  #
+  # This class is an observable responsable of periodically emitting view of the system
+  # to other components like the internal metrics pipelines.
   class Collector
-    # TODO: move to a timer based  API with concurrent ruby timer class
-    # Allow to accept an external flush
-    # When the flush is done we should not record any new metric
     include LogStash::Util::Loggable
     include Observable
     include Singleton
 
-    SNAPSHOT_ROTATION_TIME = 1 #seconds
+    SNAPSHOT_ROTATION_TIME = 1 # seconds
+    SNAPSHOT_ROTATION_TIME_TIMEOUT = 10 * 60 # seconds
 
     def initialize
-      @snapshot_rotation_mutex = Mutex.new
-      rotate_snapshot
+      @metric_store = MetricStore.new
+
+      start_periodic_snapshotting
     end
 
-    # This part of the code is called from multiple threads
-    # TODO: rename to record?
+    # The metric library will call this unique interface
+    # its the job of the collector to update the store with new metric
+    # of update the metric
+    #
+    # If there is a problem with the key or the type of metric we will record an error 
+    # but we wont stop processing events, theses errors are not considered fatal.
+    # 
     def push(*args)
-      snapshot.push(*args)
-    end
+      namespaces_path, key, type, other = args
 
-    def self.snapshot_rotation_time=(time)
-      @snapsho_rotation_time = ime
+      begin
+        metric = @metric_store.fetch_or_store(namespaces_path, key) do
+          LogStash::Instrument::MetricType.create(type, namespaces_path, key)
+        end
+        metric.execute(*other)
+        changed # we had changes coming in so we can notify the observers
+      rescue MetricStore::ConcurrentMapExpectedError => e
+        logger.error("Collector: Cannot record metric", :exception => e)
+      rescue NameError => e
+        logger.error("Collector: Cannot create concrete class for this metric type",
+                     :type => type,
+                     :namespaces_path => namespaces_path,
+                     :key => key,
+                     :stacktrace => e.backtrace)
+      end
     end
 
-    def self.snapshot_rotation_time
-      @snapshot_rotation_time || SNAPSHOT_ROTATION_TIME
+    # Monitor the `Concurrent::TimerTask` this update is triggered on every successful or not
+    # run of the task, this method will record the execution in the log
+    #
+    # @param [Time] Time of execution
+    # @param [result] Result of the execution
+    # @param [Exception] Exception
+    def update(time_of_execution, result, exception)
```

Comment:
See this line: https://github.com/elastic/logstash/pull/4358/files#diff-b949194a3f253e60f88091a1ff603dd1R86